### ISSUE TITLE
Investigate whether APT can require multiple GPG signatures

### ISSUE THREAD

**Issue by lamby** [turnkeylinux/tracker#895]
This is for the security archive.

**Comment by lamby** [comment]
Brain/linkdump:

 - https://wiki.debian.org/SecureApt

**Comment by lamby** [comment]
https://superuser.com/a/634385/95437

Possibly look at https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing

**Comment by lamby** [comment]
@JedMeister 

Asked a few people at DebConf about this. Whilst it would be theoretically possible to modify APT to require multiple signatures, the changes would be rather invasive, slow (due to extra, cautious and conservative review of anything signing/security related) and require some kind of rules tables and new syntax etc.

The other practical issue is that this would not be introduced into ``stretch`` and thus could not be relied upon to be available. Hypothetically it may not be relied upon until ``buster+1`` as upgrades to a hypothetical ``buster`` archive would not be supported by the version performing the upgrade (ie. the version in ``stretch``).

As an alternative, what I suggest is that that — if you want to persue this — to use some variant of Shamir's Secret Sharing.

The basic idea is you can split a secret (in this case, say, the private Turkney security key) into *n* parts. You can even decide that the key can be constructed from *n - 1* parts, which is somewhat sensible from a "bus factor" point of view.

There is a dummy app in Debian called ``ssss`` which demos this  if you want to play with the concept, and more stable/reliable tools and libraries in the form of ``libgfshare``.

When you would do a security update you would assemble sufficient parties and get them to provide their "parts" key. Alternatively, they could provide files on a physical USB stick, but the concept is identical and a little bit more unwieldy. The underlying GPG secret key would then be unlocked and the security archive would be signed with said, single key - everything else works just as before, we are just a) signing manually and b) not keeping the private key lying around but spreading it around *n* people.

(Oh, if you didn't want to be able to construct the key from *n - 1* parts and passphrases only then this is mostly equivalent to having a GPG key where people in a defined order enter their passwords without pressing return!)

Anyway, does this appeal at all? How are you signing the archive at the moment?

Would be neat to have a tool that could split and combine GPG keys for signing, and am excited if I could write it! Let me know :)

**Comment by JedMeister** [comment]
Thanks @lamby.

That certainly does appeal to me and a tool to do that would be very cool! How about we wait until @alonswartz is back from his business trip to US (all this week) and I'll discuss with him. Personally I think that we should make this a priority and implement it for our v15.0 (Stretch based) release.

**Comment by lamby** [comment]
Nod nod!

**Comment by lamby** [comment]
@JedMeister How we doing on this? :)

**Comment by JedMeister** [comment]
Sorry for slow response @lamby - Alon only flew back into Israel yesterday (my time). I'll be speaking with him tonight (my time). I plan to cover this in our chat.

**Comment by lamby** [comment]
@JedMeister Friendly ping on this :)

**Comment by JedMeister** [comment]
Cheers @lamby - sorry about the delay... Been crazy busy and kept forgetting to bring it up...

I've sent Alon a bump to have a look, so hopefully he'll be along shortly... :smile:

**Comment by lamby** [comment]
@JedMeister Another friendly ping on this o/ :)

**Comment by JedMeister** [comment]
Hey @lamby - apologies on the slow turnaround on this... I finally spoke with Alon about this recently (but only just posting now).

To actually answer your previous question, currently Alon manages everything, from building the pacakge, signing it and uploading it to the repo. Beyond simply hardening security, part of the rationale for this is that I will eventually taking over the package building and publishing and we wanted to implement some additional security, especially for the "security" repo.

As per usual, we've got way more bogged down than we would have liked, so are pulling back a little on a few things (unfortunately, it's a bit of a pattern for us; explore all of the cool shit we could be doing; get excited about the possibilities; then reality catches up and crashes the party, then we pull our collective heads in).

So unfortunately for now, we're going to put this project on ice. In the longer term, we're keen to revisit it. But in the short term, we need to focus our resources on getting it "good enough" rather than "ideal".

So at this point we are planning to just use separate keys for the "main" and "security" repos. I'll hold the key for "main" and Alon will continue to hold the key for "security". The down side is that we'll still need to rely on Alon to do security package builds, but TBH, we've only very rarely used it.

**Comment by lamby** [comment]
Aw, I understand :)   (Shame though; was going to be a fun project!)

**Comment by JedMeister** [comment]
Yeah it would be cool, but we'll need to leave it for another day...
